bmcs/src/main/scala/akka/stream/alpakka/oracle/bmcs/impl/BmcsStream.scala

```diff
@@ -0,0 +1,302 @@
+   *
+   * @param objectName  the object to download.
+   * @param bucket      the bucket
+   * @param rangeOption optional range
+   * @return a Future of HttpResponse that download the object from bmcs.
+   */
+  def request(bucket: String, objectName: String, rangeOption: Option[ByteRange] = None): Future[HttpResponse] = {
+    val downloadRequest: HttpRequest = getDownloadRequest(bucket, objectName)
+    signAndGet(rangeOption match {
+      case Some(range) => downloadRequest.withHeaders(headers.Range(range))
+      case _ => downloadRequest
+    })
+  }
+
+  /**
+   * Uploads a stream of ByteStrings to a specified location as a multipart upload.
+   */
+  def multipartUpload(bucket: String,
+                      objectName: String,
+                      chunkSize: Int = MinChunkSize,
+                      chunkingParallelism: Int = 4): Sink[ByteString, Future[CompleteMultipartUploadResult]] =
+    chunkAndRequest(bucket, objectName, chunkSize)(chunkingParallelism)
+      .toMat(completionSink(bucket, objectName))(Keep.right)
+
+  /**
+   * Initiates the multipart upload.
+   * https://docs.us-phoenix-1.oraclecloud.com/api/#/en/objectstorage/20160918/MultipartUpload/CreateMultipartUpload
+   *
+   * @param objectName the object that's being uploaded as multipart.
+   * @param bucket     the bucket to upload to
+   * @return The uploadId wrapped in [MultipartUpload]
+   */
+  private def initiateMultipartUpload(bucket: String, objectName: String): Future[MultipartUpload] = {
+    import mat.executionContext
+    Marshal(CreateMultipartUploadDetails(objectName))
+      .to[RequestEntity]
+      .flatMap(
+        entity =>
+          signAndGetAs[MultipartUpload](
+            initiateMultipartUploadRequest(bucket)
+              .withEntity(entity)
+        )
+      )
+  }
+
+  private def chunkAndRequest(bucket: String, objectName: String, chunkSize: Int = MinChunkSize)(
+      parallelism: Int = 4
+  ): Flow[ByteString, PartUploadResponse, NotUsed] = {
+
+    // Multipart upload requests (except for the completion api) are created here.
+    //  The initial upload request gets executed within this function as well.
+    //  The individual upload part requests are created.
+    val requestFlow: Flow[ByteString, (HttpRequest, (MultipartUpload, Int)), NotUsed] =
+      createMultipartRequests(bucket, objectName, chunkSize, parallelism)
+
+    // The individual upload part requests are processed here
+    requestFlow.via(Http().superPool[(MultipartUpload, Int)]()).map {
+      case (Success(r), (upload, index)) =>
+        r.entity.dataBytes.runWith(Sink.ignore)
+        val etag = r.headers.find(_.lowercaseName() == "etag").map(_.value)
+        etag
+          .map((t) => SuccessfulPartUpload(upload, index, t))
+          .getOrElse(FailedPartUpload(upload, index, BmcsException(r, "Cannot find etag")))
+
+      case (Failure(e), (upload, index)) => FailedPartUpload(upload, index, e)
```

Comment:
If one part fails to upload, should it really continue with the rest?

Reply:
Personally, i would want to retry in this situation. But re-try will depend on the max materializations we have configured initially. I didn't implement the retry, cause i didn't want to leave the possibility of infinite materializations to get the hash and then fail. 

Instead of collecting failures in the sink stage.. what else can i do? can i just throw an exception?

Reply:
I think it should fail the upload and let the user handle retry. But maybe that is problematic and will leave garbage on bmcs?